### ISSUE TITLE
MapPath problem with nested entitys and JPA / Hibernate 

### ISSUE THREAD

**Issue by Kernle32DLL** [querydsl/querydsl#1624]
Good morning,

I toyed around with querydsl the past few days, and ran into a strange problem with a Map<String, String>. Not sure if this is an error on querydsls or my side, so I'm posting it here.

Hibernate: 4.3.11.Final
Querydsl: 3.6.8

I have my entity set up as follows:

```java
@Entity
public class EvaluationCentreContext {
  [...]

  @ElementCollection(fetch=FetchType.EAGER)
  @CollectionTable(name = "evaluationcentrecontext_attribute",
      joinColumns = {@JoinColumn(name = "context_id")})
  @MapKeyColumn(name = "attribute_key")
  @Column(name = "attribute")
  private Map<String, String> attributes;

  [...]
}
```

The attributes map is backed by an MSSQL table as follows:

![clipboard01](https://cloud.githubusercontent.com/assets/1338408/10844257/8765a612-7efe-11e5-88e9-f5652ed1b95a.png)

The Q... class is generated correctly, too:

```java
@Generated("com.mysema.query.codegen.EntitySerializer")
public class QEvaluationCentreContext extends EntityPathBase<EvaluationCentreContext> {
    [...]
    public final MapPath<String, String, StringPath> attributes = this.<String, String, StringPath>createMap("attributes", String.class, String.class, StringPath.class);
    [...]
}
```

Now, I want to order my EvaluationCentreContext objects via an attribute in the attributes map. This works as intendend:

```java
evaluationCentreContextRepository.findAll(null, QEvaluationCentreContext.evaluationCentreContext.attributes.get("bla123").asc());
```

Which results in

```sql
select
	[...]
from
	evaluationcentrecontext evaluation0_ 
inner join
	evaluationcentrecontext_attribute attributes1_ 
		on evaluation0_.auto_id=attributes1_.context_id 
		and (
			attributes1_.attribute_key='bla123'
		) 
order by
	attributes1_.attribute asc
```

However, as soon as I introduce a new entity, that has an reference to said EvaluationCentreContex, the query breaks...

```java
@Entity
public class RadiationExposedActivity{
  [...]


  @OneToOne(optional = false)
  @JoinColumn(name = "context_id")
  private EvaluationCentreContext evaluationCentreContext;
    [...]
}
```

```java
@Generated("com.mysema.query.codegen.EntitySerializer")
public class QRadiationExposedActivity extends EntityPathBase<RadiationExposedActivity> {
    [...]
        public final QEvaluationCentreContext evaluationCentreContext;
        [...]
}
```

```java
QRadiationExposedActivity qradiationExposedActivity = QRadiationExposedActivity.radiationExposedActivity;
radiationExposedActivityRepository.findAll(null, qradiationExposedActivity.evaluationCentreContext.attributes.get("bla123").asc());
```

```sql
select
	[...]
from
	radiationexposedactivity radiatione0_ 
inner join
	evaluationcentrecontext evaluation1_ 
		on radiatione0_.context_id=evaluation1_.auto_id 
		and (
			attributes2_.attribute_key='bla123'
		) 
inner join
	evaluationcentrecontext_attribute attributes2_ 
		on evaluation1_.auto_id=attributes2_.context_id 
order by
	attributes2_.attribute asc
```

Error message is: The multi-part identifier "attributes2_.attribute_key" could not be bound.

The error is obvious - the "and" part in the first inner join does not belong there. It should be part of the evaluationcentrecontext_attribute join. If I manually edit the query, it works ofc.

So - where did I go wrong? Is this a problem with querydsl, is hibernate messing up the query, or is this a stupid mistake on my end?

Link to Google Group: https://groups.google.com/forum/#!topic/querydsl/fPMNqQnkhaI

Greetings,
Björn

**Comment by Kernle32DLL** [comment]
After some debugging I don't think this is Querydsl related. Looks like hibernate :/

I analyzed the native query Querydsl generates... (shortened)

```jpql
select rep from radiationexposedactivity rep inner join rep.evaluationCentreContext.attributes as evalAttribute with key(evalAttribute) = 'bla123'
```

That looks pretty good, and should work from my understanding. But it doesn't. Same error message.

Suggestions?

**Comment by timowest** [comment]
I guess we need to split the join into two parts to make it work:

```
select rep from radiationexposedactivity rep inner join rep.evaluationCentreContext.attributes as evalAttribute with key(evalAttribute) = 'bla123'
```

becomes

```
select rep from radiationexposedactivity rep 
inner join rep.evaluationCentreContext as evaluationCentreContext
inner join evaluationCentreContext.attributes as evalAttribute with key(evalAttribute) = 'bla123'
```

**Comment by timowest** [comment]
I will do that change in the PR I prepared.

**Comment by Kernle32DLL** [comment]
Thanks for fixing this (and so fast!).

Could this be backported to 3.X? We are forced to use 3.X due to the incompatibility between 4.X and hibernate :/.

**Comment by timowest** [comment]
Yes, we can backport it. What kind of Hibernate issues do you have with Querydsl 4?

**Comment by Kernle32DLL** [comment]
Don't remeber if it was a hibernate or spring issue. But the problem is that affected framework did not change the packagename, and thus not accepts 4.X classes.

**Comment by Kernle32DLL** [comment]
My bad. Its Spring Data causing the problem.

https://jira.spring.io/browse/DATACMNS-670

I did read different things about wheater they are going to fix it or drop support for querydsl.

**Comment by Kernle32DLL** [comment]
So how is the backport going? I saw 3 failed checks? :/

**Comment by timowest** [comment]
The backport is now fixed, some other changes needed to be backported as well. Once it is merged you can access it from the nightly snapshots or wait for the next release.

**Comment by Kernle32DLL** [comment]
Where am I able to obtain said nightly snapshots? Looking at past release dates of QDSL I assume that the next release will be at the end of the month? If so, we depend on the nightlys, since we are scheduled for release next week.

Don't mean to rush you tho. We really appreciate your work and support.

**Comment by timowest** [comment]
The snapshots repository is oss.sonatype.org/content/repositories/snapshots

I'll push the review of the PR.

**Comment by timowest** [comment]
It's now merged and should be available in the next nightly snapshot.

**Comment by Kernle32DLL** [comment]
Fix confirmed with the 3.7.0 Build Snapshot 20151108. Thanks for your support! :)
